quark/core/r2apkinfo.py

```diff
@@ -0,0 +1,739 @@
+# -*- coding: utf-8 -*-
+# This file is part of Quark-Engine - https://github.com/quark-engine/quark-engine
+# See the file 'LICENSE' for copying permission.
+
+import functools
+import logging
+import os.path
+import re
+import tempfile
+import zipfile
+from collections import defaultdict, namedtuple
+from os import PathLike
+from typing import Any, Dict, Generator, List, Optional, Set, Tuple, Union
+
+import r2pipe
+
+from quark.core.axmlreader import AxmlReader
+from quark.core.interface.baseapkinfo import BaseApkinfo, XMLElement
+from quark.core.struct.bytecodeobject import BytecodeObject
+from quark.core.struct.methodobject import MethodObject
+from quark.utils.tools import (
+    descriptor_to_androguard_format,
+    remove_dup_list,
+)
+
+R2Cache = namedtuple("r2_cache", "address is_imported")
+
+PRIMITIVE_TYPE_MAPPING = {
+    "void": "V",
+    "boolean": "Z",
+    "byte": "B",
+    "char": "C",
+    "short": "S",
+    "int": "I",
+    "long": "J",
+    "float": "F",
+    "double": "D",
+}
+
+R2_ESCAPE_CHAR_LIST = ["$"]
+
+
+class R2Imp(BaseApkinfo):
+    def __init__(
+        self,
+        apk_filepath: Union[str, PathLike],
+        tmp_dir: Union[str, PathLike] = None,
+    ):
+        super().__init__(apk_filepath, "r2")
+
+        if self.ret_type == "DEX":
+            self._tmp_dir = None
+            self._dex_list = [apk_filepath]
+
+        elif self.ret_type == "APK":
+            self._tmp_dir = tempfile.mkdtemp() if tmp_dir is None else tmp_dir
+
+            with zipfile.ZipFile(self.apk_filepath) as apk:
+                apk.extract("AndroidManifest.xml", path=self._tmp_dir)
+
+                self._manifest = os.path.join(self._tmp_dir, "AndroidManifest.xml")
+
+                dex_files = [
+                    file
+                    for file in apk.namelist()
+                    if file.startswith("classes") and file.endswith(".dex")
+                ]
+
+                for dex in dex_files:
+                    apk.extract(dex, path=self._tmp_dir)
+
+                self._dex_list = [os.path.join(self._tmp_dir, dex) for dex in dex_files]
```

Comment:
I suggest removing these (Line 58 - 72) and letting R2 handle the multi-dex issue.

Reply:
Remove `self._dex_index` .

Reply:
What's the reason not to remove `self._dex_list`? It's a private variable and not used by other code in `r2apkinfo.py`

Reply:
Done.